OpenSim/Sandbox/DataFormattersAndAdapters.h

```diff
@@ -0,0 +1,212 @@
+public:
+    MotDataStreamFormatter();
+
+private:
+    unique_ptr<DataStreamFormatter> implementClone() const
+    {   return MotDataStreamFormatter(*this); }
+
+    OutputTables implementRead(std::istream& motIn) const override;
+    void implementWrite(std::ostream& motOut, InputTables) override;
+};
+
+//==============================================================================
+//                         C3D FORMATTER
+//==============================================================================
+/** A standalone wrapper class for the BTK library; this is not a 
+DataStreamFormatter. **/
+class C3dFormatter {
+public:
+    // Useful stuff for dealing with BTK.
+};
+
+//==============================================================================
+//                       DATA ADAPTER classes
+//==============================================================================
+/** Associated with a data source/sink of a known type and can read and write
+DataTables from/to that source/sink. 
+
+Note that these objects are associated with both a format and a data source,
+which might be a file or cin or a string or whatever.
+**/
+class DataAdapter {
+public:
+    OutputTables read() const;
+    void write(InputTables);
+    virtual ~DataAdapter() = default;
+
+protected:
+    DataAdapter() = default;
+    virtual OutputTables implementRead() const = 0;
+    virtual void implementWrite(InputTables) = 0;
+};
+
+/** This is a DataAdapter for use with an already-existing read/write stream.
+For read-only streams (istream) like std::cin or write-only (ostream) like
+std::cout we'll need IStreamDataAdapter and OStreamDataAdapter also. Those
+should raise exceptions if disallowed read() or write() is called.
+**/
+class StreamDataAdapter : public DataAdapter {
+public:
+    /** Associate this adapter with a stream and a format for that stream. It
+    is an error if the formatKey cannot be found in the DataStreamFormatter
+    registry. **/
+    StreamDataAdapter(iostream& stream, const std::string& formatKey) 
+    :   m_stream(stream) {
+        m_formatter = DataStreamFormatter::createFormatter(formatKey);
+    }
+private:
+    OutputTables implementRead() const override 
+    {   return m_formatter->read(m_stream); }
+    void implementWrite(InputTables input) override 
+    {   return m_formatter->write(m_stream, input); }
+
+    std::iostream&                  m_stream;
+    unique_ptr<DataStreamFormatter> m_formatter;
+};
+
+/** A string adapter is a kind of stream adapter. **/
+class StringDataAdapter : public StreamDataAdapter {
+public:
+    StringDataAdapter(const std::string& data, const std::string& formatKey)
+    :   StreamDataAdapter(std::stringstream(data), formatKey) {}
+};
+
+/** Helper class for when you know this file will be handled using one of
+the DataStreamFormatters. Users will use the FileDataAdapter instead. **/
+class FileStreamDataAdapter : public StreamDataAdapter {
```

Comment:
> It seems like there is a distinction between plain text files and non-plain-text files. Is this true? 

No, there should be no distinction. The particular format implies whether the expected stream will be treated as binary or text. But otherwise they are both just streams of bytes. In code, the distinction occurs only at the time a file is opened, and the only significance is that line ending characters are meaningful in text files but not in binary files.